// @identity-connect/wallet-api/src/serialization/bcsSerialization.ts
// Copyright © Aptos
// SPDX-License-Identifier: Apache-2.0

import { Deserializer, Hex, Serializable } from '@aptos-labs/ts-sdk';

export type BcsDeserializableV2Class<T extends Serializable> = {
  deserialize(deserializer: Deserializer): T;
};

/**
 * Check if a value is BCS serializable
 */
export function isBcsSerializable(value: any): value is Serializable {
  return (
    (value as Serializable)?.serialize !== undefined &&
    (value as Serializable)?.bcsToBytes !== undefined &&
    (value as Serializable)?.bcsToHex !== undefined
  );
}

export function bcsSerialize(serializable: Serializable) {
  return serializable.bcsToHex().toString();
}

export function bcsDeserialize<T extends Serializable>(
  deserializableClass: BcsDeserializableV2Class<T>,
  serializedValue: string,
) {
  const serializedValueBytes =
    Hex.fromHexString(serializedValue).toUint8Array();
  const deserializer = new Deserializer(serializedValueBytes);
  return deserializableClass.deserialize(deserializer);
}
